---
title: "Untitled"
author: "Ashley Tseng"
date: "10/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Download external and command line packages
Follow the instructions to download and install:

* [MAFFT](https://mafft.cbrc.jp/alignment/software/)
* Download [IQTREE](http://www.iqtree.org/) and save within the "Phylo" folder.

## Load data
load a fasta file:
```{r}
fastafile = read.fasta(file = paste0( path, "variants.fasta"))
class(fastafile)
```

load meta data: 
```{r}
metadata = read.csv(file = paste0( path, "metadata.csv"))
class(metadata)
```



# Alignments
## MAFFT: is the most updated sequence alignment program; very efficient

We will use MAFFT (FFT-NS-2) to align these sequences. MAFFT is efficient for very very large alignments. Read more about MAFFT here: https://mafft.cbrc.jp/alignment/software/

There are two ways you can run this algorithm. First, you can run this command line using the following code: 

**For Mac:**
```
cd "/Users/ashleytseng/OneDrive - UW/UW PhD/Jobs/Phylogenetics/Lab Group/Diana Phylogenetics Exercise_12.02.2020/phylogenetics_example/Phylo/"
mafft --thread -1 --nomemsave sequences.fasta > alignment.fasta
```
# "thread -1" tells the program to use as many cores as possible to run
# "nomemsave" tells the program not to save to memory
# "alignment.fasta" tells the program what we want to name the output

**For Windows**
```
cd /d \Users\dianatordoff/Desktop\Phylo\
mafft --thread -1 --nomemsave sequences.fasta > alignment.fasta
```

Or you can run it in R using the following code:
```{r}
system("mafft --thread -1 --nomemsave sequences.fasta > MAFFTalignment.fasta")
```

## Load and View Alignments
Now we load the alignment:
```{r}
#as a list
alignment <- read.fasta(file = paste0( path, "MAFFTalignment.fasta"))
class(alignment)

#as a DNA string set
dna <- readAAStringSet(paste0( path, "MAFFTalignment.fasta"))
class(dna)
```

And view them...

View first sequence:
```{r}
alignment[1]
```

View alignment object using ``ggmsa`` (see [vignette for details](https://cran.r-project.org/web/packages/ggmsa/vignettes/ggmsa.html) ):
```{r}
#start of alignemnt
ggmsa(dna[1:50], start=0, end=100)

#end of alignment
l=length(dna$seq1)
ggmsa(dna[1:50], start=l-250, end=l)
ggmsa(dna[1:50], start=l-500, end=l-250)

#view full alignment (takes several minutes):
#ggmsa(dna, font = NULL)
```


**NOTE:** the ``msa`` package has options to align sequences using Clustal and MUSCLE. See more information [here](https://bioconductor.org/packages/release/bioc/vignettes/msa/inst/doc/msa.pdf)



# Editing fasta files
## Trimming

Examine sequence lengths:
```{r}
length <- getLength(fastafile)
hist(length, type="density")
summary(length) 
```

Based on the above view of the alignment, we will trim the leading 10 and trailing 400 bp from the alignment. 
```{r}
l=length(dna$seq1)
trimmed <- alignment  
for (i in 1:500) {
  trimmed[[i]] <- trimmed[[i]][10:(l-400)]
}
length(trimmed$seq1)
```

Save trimmed alignment:
```{r}
write.fasta(trimmed, names=names(trimmed), file.out=paste0( path, "trimmed_alignment.fasta"))
```

## Creating sequence subsets

Suppose we want a subset of sequences from MSM/IDU only:
```{r}
table(metadata$transm)

# I want to keep all sequence names of the seqeunces that are of transmission category "MSM/IDU" in the metadata
keepnames = metadata[metadata$transm == "MSM/IDU",]$seqID
trimmed.MSMIDUonly <- trimmed[ c(which( names(trimmed) %in% keepnames )) ]
# We can also do the above in one set
trimmed.MSMIDUonly <- trimmed[ c(which( names(trimmed) %in% metadata[metadata$transm=="MSM/IDU",]$seqID )) ]

write.fasta(trimmed.MSMIDUonly, names=names(trimmed.MSMIDUonly), file.out=paste0( path, "trimmed.MSMIDUonly_alignment.fasta"))
```





# Reconstructing Phylogeny

## Neighbor Joining: clustering method for a pairwise distance matrix; "upgma" is an alternative clustering method

Here is a simple example of a neighbor joining tree of the MSM/IDU subset:

need to load as a dnabin object:
```{r}
dna.MSMIDU <- read.dna(file = paste0( path, "trimmed.MSMIDUonly_alignment.fasta"), format="fasta")
class(dna.MSMIDU)
```

```{r}
tree.MSMIDU <- njs(dist.dna(dna.MSMIDU))
plot(tree.MSMIDU)
```


## IQTREE

IQTREE is very efficient for large datasets. You do need to specify your own substitution model, however. IQTREE is a maximum likelihood estimation.

Go [here](http://www.iqtree.org/doc/Substitution-Models#dna-models) and [here](http://www.iqtree.org/doc/Substitution-Models#rate-heterogeneity-across-sites) to learn more about substitution models. All of these are ways to adjust branch lengths to account for saturation (the number of multiple hits over time that can obscure differences).

Rate heterogeneity is trying to account for the fact that some sites in a gene evolves a lot more quickly than other sites. Uses likelihood/probabilities of change to account for some sites where there are a lot of observed differences and not so much for other sites. You're not telling the algorithm which sites you're expecting a lot of differences in (does not assume prior knowledge).

**For Mac:**
```
cd "/Users/ashleytseng/OneDrive - UW/UW PhD/Jobs/Phylogenetics/Lab Group/Diana Phylogenetics Exercise_12.02.2020/phylogenetics_example/Phylo/"
iqtree-1.6.12-MacOSX/bin/iqtree -s trimmed_alignment.fasta -m HKY+G4+I+R -nt 2
```
"iqtree -s" indicates the sequences you're using
"-m" specifies the substitution model you're using 
"-nt" specifies the number of threads you're using 

There's three phases IQTREE goes through:
1. Reads in the alignment
2. Initializing candidate tree set
3. Optimizing candidate tree set

IQTREE saves a bunch of files, but these are the 2 important ones:
1. "trimmed_alignment.fasta.log" saves everything in the terminal; lets you look back at the code
2. "trimmed_alignment.fasta.treefile": is your actual tree

**For Windows**
```
cd /d \Users\dianatordoff\Desktop\Phylo\
iqtree-1.6.12-Windows\bin\iqtree -s trimmed_alignment.fasta -m HKY+G4+I+R -nt 2
```

*NOTE: -nt 2 indicates the number of threads, or cores, for parallel computing. Most laptops have 2, but you can increase this if you have additional processors*

Or you can run it in R using the following code:
```{r}
system("iqtree-1.6.12-MacOSX/bin/iqtree -s trimmed_alignment.fasta -m HKY+G4+I+R -nt 2")
```






# Plotting Trees

Load tree:
```{r}
setwd("/Users/ashleytseng/phylip-3.695/exe")
tree = read.newick("outtree")
ggtree(tree)
ggtree(tree, layout="slanted") 
ggtree(tree, layout="circular")
ggtree(tree, layout="fan", open.angle=120)
ggtree(tree, layout="equal_angle")
ggtree(tree, branch.length='none')
ggtree(tree, branch.length='none', layout='circular')
```

## root tree

```{r}
tree = root(tree, "Reference") 
is.rooted(tree)
```

## extract a subclade

```{r}
tree$Nnode
subtree = extract.clade(tree, 3)
Ntip(subtree)
```

## ggtree

This is a great reference guide for using [ggtree](https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html#introduction-1).

There are several ways to plot a tree:
```{r}
ggtree(subtree)
ggtree(subtree, layout="slanted") 
ggtree(subtree, layout="circular")
ggtree(subtree, layout="fan", open.angle=120)
ggtree(subtree, layout="equal_angle")
ggtree(subtree, branch.length='none')
ggtree(subtree, branch.length='none', layout='circular')
```

You can also attach metadata (this is one of several ways to attach metadata):
```{r}
df <- tibble(seqID = tree$tip.label)
df <- merge(df, metadata, by="seqID", all.x=TRUE)

p <- ggtree(tree, branch.length='none', layout='circular')
p %<+% df + geom_tippoint(aes(color = transm))

q <- ggtree(tree)
q %<+% df + geom_tippoint(aes(color = age)) + 
  scale_colour_gradient2(low="orange", mid="white", high="blue", midpoint=35)
```


# Genetic Distances & Clusters

Here we use a simplification of Art Poon's patristic distance method R function from "ape" package.
```{r}
#define function (uses igraph and network graphing)
cluster_by_patristic_distance <- function(tree, min_size = 2, threshold = 0.02) {
    patristic.dist <- ape::cophenetic.phylo(tree)
    adj_mat        <- patristic.dist <= threshold
    cluster_graph  <- igraph::graph_from_adjacency_matrix(adj_mat, mode="undirected", diag=FALSE)
    components     <- igraph::components(cluster_graph)
    retain_assignments <- purrr::keep(components$membership, ~ components$csize[.x] >= min_size)
    clusters           <- tibble(sequence_name = names(retain_assignments), cluster = retain_assignments)
  return(clusters)
}

patristic.02.sarscov2 <- cluster_by_patristic_distance(tree, min_size = 2, threshold = 0.02)
```

Here we use Tamura-Nei (TN93) pairwise distance, which is equivalent to HIV-TRACE. R function from "ape" package. 
```{r}
dna <- read.dna(file = paste0( path, "trimmed_alignment.fasta"), format="fasta")
class(dna) #needs to be a DNAbin object

cluster_by_TN93 <- function(tree, min_size = 2, threshold = 0.02) {
    TN93.dist     <- ape::dist.dna(dna, model="TN93", pairwise.deletion = TRUE, as.matrix=TRUE)
    adj_mat       <- TN93.dist <= threshold
    cluster_graph <- igraph::graph_from_adjacency_matrix(adj_mat, mode="undirected", diag=FALSE)
    components    <- igraph::components(cluster_graph)
    retain_assignments <- purrr::keep(components$membership, ~ components$csize[.x] >= min_size)
    clusters           <- tibble::tibble(sequence_name = names(retain_assignments), cluster = retain_assignments)
  return(clusters)
}

tn93.02.hiv <- cluster_by_TN93(dna, min_size = 2, threshold = 0.02)
```
